hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/RMAppManager.java

```
@@ -467,28 +474,33 @@ private RMAppImpl createAndPopulateNewRMApp(
       if (scheduler instanceof CapacityScheduler) {
         String queueName = placementContext == null ?
             submissionContext.getQueue() : placementContext.getFullQueuePath();
-
-        String appName = submissionContext.getApplicationName();
-        CSQueue csqueue = ((CapacityScheduler) scheduler).getQueue(queueName);
-
-        if (csqueue == null && placementContext != null) {
-          //could be an auto created queue through queue mapping. Validate
-          // parent queue exists and has valid acls
-          String parentQueueName = placementContext.getParentQueue();
-          csqueue = ((CapacityScheduler) scheduler).getQueue(parentQueueName);
+        CapacityScheduler cs = (CapacityScheduler) scheduler;
+        CSQueue csqueue = cs.getQueue(queueName);
+        PrivilegedEntity privilegedEntity = getPrivilegedEntity(
+            csqueue == null ? queueName : csqueue.getQueuePath());
+
+        YarnAuthorizationProvider dynamicAuthorizer = null;
+        boolean ambiguous = false;
+        if (csqueue == null) {
+          try {
+            List<Permission> permissions =
+                cs.getCapacitySchedulerQueueManager().getPermissionsForDynamicQueue(
+                    new QueuePath(queueName), cs.getConfiguration());
+            if (!permissions.isEmpty()) {
+              dynamicAuthorizer = new ConfiguredYarnAuthorizer();
+              dynamicAuthorizer.setPermission(permissions, userUgi);
+            }
+          } catch (SchedulerDynamicEditException e) {
+            ambiguous = true;
```

getPermissionForDynamicQueue could produce a few exceptions other than ambiguous queue. It is also silently swallowing the exception. I suggest a renaming and a logging mechanism here.

---

Unfortunately determineMissingParents was not really designed for reusability in mind, and I didn't want to refactor it, but, I tried to introduce different types of exceptions for different errors, e.g: SchedulerDynamicMissingParentException, SchedulerDynamicAmbiguousParentException, SchedulerDynamicAboveMaxDepthLimitException etc.

This whole ambiguous thing is a bit misleading, two test cases:

1. TestCapacitySchedulerAmbiguousLeafs.testAmbiguousSubmissionWithACL: the queue is ambiguous not the parent, so the determineMissingParents would throw SchedulerDynamicMissingParentException, for ambiguity the queue must be checked beforehand. The app submission will fail later if we don't check queue acls here.

2. TestApplicationACLs.verifyInvalidQueueWithAcl: the queue is invalid, and no acl checks should be done on it. The app submission will fail later.


The conclusion is that we must ONLY check dynamic queue ACLs when we can construct the dynamic missing parents. The error logging is not relevant, the app submission will fail anyway (hopefully in every case, I didn't touch that logic with this PR..., to be honest it looks fragile that in some "error" like cases we don't check queue ACLs).


Removed the ambiguous variable, and now the code is much cleaner:

```
        YarnAuthorizationProvider dynamicAuthorizer = null;
        if (csqueue == null) {
          try {
            List<Permission> permissions =
                cs.getCapacitySchedulerQueueManager().getPermissionsForDynamicQueue(
                    new QueuePath(queueName), cs.getConfiguration());
            if (!permissions.isEmpty()) {
              dynamicAuthorizer = new ConfiguredYarnAuthorizer();
              dynamicAuthorizer.setPermission(permissions, userUgi);
            }
          } catch (SchedulerDynamicEditException ignored) {
          }
        }

        if (csqueue != null || dynamicAuthorizer != null) {
           // here we have to check the queue ACLs
```

---

I think a cleaner solution would be to catch the exception in getPermissionForDynamicQueue, and return an empty permission list in case of an error. I would add at least a debug log for the error just in case. It is rarely a good idea to ignore an exception. I do think it would be useful to see why ACLs are ignored for a dynamic queue.

---

Done.